File: src/main/java/com/matchbox/book/OrderBook.java
package com.matchbox.book;

import com.matchbox.core.*;
import java.util.*;

public class OrderBook {

    private final TreeMap<Long, PriceLevel> bids;
    private final TreeMap<Long, PriceLevel> asks;
    private final Map<Long, Order> orderMap;
    private final List<Trade> tradeLog;

    public OrderBook() {
        this.bids = new TreeMap<>(Comparator.reverseOrder());
        this.asks = new TreeMap<>(Long::compareTo);
        this.orderMap = new HashMap<>();
        this.tradeLog = new ArrayList<>();
    }

    public void processOrder(OrderCommand cmd) {
        switch (cmd.getType()) {
            case LIMIT -> processLimit(cmd.getOrder());
            case MARKET -> processMarket(cmd.getOrder());
            case CANCEL -> processCancel(cmd.getCancelOrderId());
            case MODIFY -> processModify(cmd.getCancelOrderId(),
                    cmd.getModifyPrice(), cmd.getModifyQuantity());
        }
    }

    private void processLimit(Order taker) {
        if (taker.getTimeInForce() == TimeInForce.FOK) {
            long available = taker.getSide() == Side.BUY
                    ? totalAskQuantityUpTo(taker.getPrice())
                    : totalBidQuantityDownTo(taker.getPrice());
            if (available < taker.getQuantity()) return;
        }

        if (taker.getSide() == Side.BUY) {
            matchBids(taker, true);
        } else {
            matchAsks(taker, true);
        }

        if (!taker.isFilled()) {
            if (taker.getTimeInForce() == TimeInForce.IOC
                    || taker.getTimeInForce() == TimeInForce.FOK) return;
            addToBook(taker);
        }
    }

    private void processMarket(Order taker) {
        if (taker.getTimeInForce() == TimeInForce.FOK) {
            long available = taker.getSide() == Side.BUY
                    ? totalAskQuantity() : totalBidQuantity();
            if (available < taker.getQuantity()) return;
        }

        if (taker.getSide() == Side.BUY) {
            matchBids(taker, false);
        } else {
            matchAsks(taker, false);
        }
    }

    private void matchBids(Order taker, boolean isLimit) {
        while (!taker.isFilled() && !asks.isEmpty()) {
            Map.Entry<Long, PriceLevel> bestAskEntry = asks.firstEntry();
            if (isLimit && taker.getPrice() < bestAskEntry.getKey()) break;

            PriceLevel level = bestAskEntry.getValue();
            matchAgainstLevel(taker, level);
            if (level.isEmpty()) asks.pollFirstEntry();
        }
    }

    private void matchAsks(Order taker, boolean isLimit) {
        while (!taker.isFilled() && !bids.isEmpty()) {
            Map.Entry<Long, PriceLevel> bestBidEntry = bids.firstEntry();
            if (isLimit && taker.getPrice() > bestBidEntry.getKey()) break;

            PriceLevel level = bestBidEntry.getValue();
            matchAgainstLevel(taker, level);
            if (level.isEmpty()) bids.pollFirstEntry();
        }
    }

    private void matchAgainstLevel(Order taker, PriceLevel level) {
        while (!taker.isFilled() && !level.isEmpty()) {
            Order resting = level.peek();
            long matchQty = Math.min(taker.getQuantity(), resting.getQuantity());
            long ts = System.nanoTime();

            tradeLog.add(new Trade(level.getPrice(), matchQty, ts,
                    resting.getOrderId(), taker.getOrderId()));

            resting.reduceQuantity(matchQty);
            taker.reduceQuantity(matchQty);

            if (resting.isFilled()) {
                level.poll();
                orderMap.remove(resting.getOrderId());
            }
        }
    }

    private void addToBook(Order order) {
        TreeMap<Long, PriceLevel> book = order.getSide() == Side.BUY ? bids : asks;
        PriceLevel level = book.get(order.getPrice());
        if (level == null) {
            level = new PriceLevel(order.getPrice());
            book.put(order.getPrice(), level);
        }
        level.add(order);
        orderMap.put(order.getOrderId(), order);
    }

    private void processCancel(long orderId) {
        Order order = orderMap.remove(orderId);
        if (order == null) return;

        TreeMap<Long, PriceLevel> book = order.getSide() == Side.BUY ? bids : asks;
        PriceLevel level = book.get(order.getPrice());
        if (level != null) {
            level.remove(order);
            if (level.isEmpty()) book.remove(order.getPrice());
        }
    }

    private void processModify(long orderId, long newPrice, long newQuantity) {
        Order existing = orderMap.get(orderId);
        if (existing == null) return;

        if (existing.getPrice() == newPrice) {
            existing.setQuantity(newQuantity);
        } else {
            Side side = existing.getSide();
            processCancel(orderId);
            processLimit(new Order(orderId, side, newPrice, newQuantity,
                    System.nanoTime(), existing.getTimeInForce(),
                    existing.getExpiryTimestamp()));
        }
    }

    // --- Order expiry ---

    public int expireOrders(long currentTimeNanos) {
        List<Long> toExpire = new ArrayList<>();
        for (Order order : orderMap.values()) {
            if (order.getTimeInForce() == TimeInForce.GTD
                    && order.getExpiryTimestamp() <= currentTimeNanos) {
                toExpire.add(order.getOrderId());
            }
        }
        for (long id : toExpire) {
            processCancel(id);
        }
        return toExpire.size();
    }

    // --- Query helpers ---

    private long totalAskQuantity() {
        long total = 0;
        for (PriceLevel level : asks.values()) {
            total += level.totalQuantity();
        }
        return total;
    }

    private long totalBidQuantity() {
        long total = 0;
        for (PriceLevel level : bids.values()) {
            total += level.totalQuantity();
        }
        return total;
    }

    private long totalAskQuantityUpTo(long price) {
        long total = 0;
        for (Map.Entry<Long, PriceLevel> e : asks.entrySet()) {
            if (e.getKey() > price) break;
            total += e.getValue().totalQuantity();
        }
        return total;
    }

    private long totalBidQuantityDownTo(long price) {
        long total = 0;
        for (Map.Entry<Long, PriceLevel> e : bids.entrySet()) {
            if (e.getKey() < price) break;
            total += e.getValue().totalQuantity();
        }
        return total;
    }

    // --- Query methods ---

    public TopOfBook getTopOfBook() {
        long bidPrice = Long.MIN_VALUE, bidQty = 0;
        long askPrice = Long.MAX_VALUE, askQty = 0;

        if (!bids.isEmpty()) {
            Map.Entry<Long, PriceLevel> e = bids.firstEntry();
            bidPrice = e.getKey();
            bidQty = e.getValue().totalQuantity();
        }
        if (!asks.isEmpty()) {
            Map.Entry<Long, PriceLevel> e = asks.firstEntry();
            askPrice = e.getKey();
            askQty = e.getValue().totalQuantity();
        }
        return new TopOfBook(bidPrice, askPrice, bidQty, askQty);
    }

    public List<BookDepthEntry> getBidDepth() {
        List<BookDepthEntry> result = new ArrayList<>();
        for (Map.Entry<Long, PriceLevel> e : bids.entrySet()) {
            result.add(new BookDepthEntry(e.getKey(), e.getValue().totalQuantity(), e.getValue().size()));
        }
        return result;
    }

    public List<BookDepthEntry> getAskDepth() {
        List<BookDepthEntry> result = new ArrayList<>();
        for (Map.Entry<Long, PriceLevel> e : asks.entrySet()) {
            result.add(new BookDepthEntry(e.getKey(), e.getValue().totalQuantity(), e.getValue().size()));
        }
        return result;
    }

    public List<Trade> getTradeLog() {
        return Collections.unmodifiableList(tradeLog);
    }

    public void clearTradeLog() {
        tradeLog.clear();
    }

    public int orderCount() {
        return orderMap.size();
    }
}

File: src/book/OrderBook.cpp
#include "matchbox/book/OrderBook.hpp"

#include <algorithm>

#include "matchbox/core/Clock.hpp"

namespace matchbox {

OrderBook::OrderBook()
    : bids_(PriceComparator{true}),   // descending: best bid = highest price
      asks_(PriceComparator{false}) {} // ascending: best ask = lowest price

void OrderBook::processOrder(const OrderCommand& cmd) {
    switch (cmd.type()) {
        case OrderType::LIMIT:
            processLimit(cmd.order());
            break;
        case OrderType::MARKET:
            processMarket(cmd.order());
            break;
        case OrderType::CANCEL:
            processCancel(cmd.cancelOrderId());
            break;
        case OrderType::MODIFY:
            processModify(cmd.cancelOrderId(), cmd.modifyPrice(), cmd.modifyQuantity());
            break;
    }
}

void OrderBook::processLimit(const std::shared_ptr<Order>& taker) {
    // Fill-Or-Kill: reject up front unless the crossable depth covers it in full.
    if (taker->timeInForce() == TimeInForce::FOK) {
        long available = taker->side() == Side::BUY
                             ? totalAskQuantityUpTo(taker->price())
                             : totalBidQuantityDownTo(taker->price());
        if (available < taker->quantity()) return;
    }

    if (taker->side() == Side::BUY) {
        matchAgainstBook(*taker, asks_, true);
    } else {
        matchAgainstBook(*taker, bids_, true);
    }

    if (!taker->isFilled()) {
        // IOC/FOK never rest: cancel whatever did not cross immediately.
        if (taker->timeInForce() == TimeInForce::IOC ||
            taker->timeInForce() == TimeInForce::FOK) {
            return;
        }
        addToBook(taker);
    }
}

void OrderBook::processMarket(const std::shared_ptr<Order>& taker) {
    if (taker->timeInForce() == TimeInForce::FOK) {
        long available = taker->side() == Side::BUY ? totalAskQuantity()
                                                    : totalBidQuantity();
        if (available < taker->quantity()) return;
    }

    if (taker->side() == Side::BUY) {
        matchAgainstBook(*taker, asks_, false);
    } else {
        matchAgainstBook(*taker, bids_, false);
    }
    // Market orders never rest: any unfilled remainder is simply dropped.
}

void OrderBook::matchAgainstBook(Order& taker, Book& book, bool isLimit) {
    while (!taker.isFilled() && !book.empty()) {
        auto best = book.begin();  // comparator guarantees this is the best price
        long bestPrice = best->first;
        if (isLimit) {
            // Buy stops once the best ask is above its limit; sell once the
            // best bid is below its limit.
            if (taker.side() == Side::BUY && taker.price() < bestPrice) break;
            if (taker.side() == Side::SELL && taker.price() > bestPrice) break;
        }
        matchAgainstLevel(taker, best->second);
        if (best->second.isEmpty()) book.erase(best);
    }
}

void OrderBook::matchAgainstLevel(Order& taker, PriceLevel& level) {
    while (!taker.isFilled() && !level.isEmpty()) {
        std::shared_ptr<Order> resting = level.peek();
        long matchQty = std::min(taker.quantity(), resting->quantity());
        long ts = nowNanos();

        tradeLog_.push_back(Trade{level.price(), matchQty, ts,
                                  resting->orderId(), taker.orderId()});

        resting->reduceQuantity(matchQty);
        taker.reduceQuantity(matchQty);

        if (resting->isFilled()) {
            level.poll();
            orderMap_.erase(resting->orderId());
        }
    }
}

void OrderBook::addToBook(const std::shared_ptr<Order>& order) {
    Book& book = order->side() == Side::BUY ? bids_ : asks_;
    auto it = book.find(order->price());
    if (it == book.end()) {
        it = book.emplace(order->price(), PriceLevel(order->price())).first;
    }
    it->second.add(order);
    orderMap_[order->orderId()] = order;
}

void OrderBook::processCancel(long orderId) {
    auto it = orderMap_.find(orderId);
    if (it == orderMap_.end()) return;

    std::shared_ptr<Order> order = it->second;
    orderMap_.erase(it);

    Book& book = order->side() == Side::BUY ? bids_ : asks_;
    auto levelIt = book.find(order->price());
    if (levelIt != book.end()) {
        levelIt->second.remove(order);
        if (levelIt->second.isEmpty()) book.erase(levelIt);
    }
}

void OrderBook::processModify(long orderId, long newPrice, long newQuantity) {
    auto it = orderMap_.find(orderId);
    if (it == orderMap_.end()) return;
    std::shared_ptr<Order> existing = it->second;

    if (existing->price() == newPrice) {
        // Same price: resize in place, keeping the order's time priority.
        existing->setQuantity(newQuantity);
    } else {
        // New price: cancel-replace. The re-inserted order loses time
        // priority (fresh entry time) and may cross on its way in.
        Side side = existing->side();
        TimeInForce tif = existing->timeInForce();
        long expiry = existing->expiryTimestamp();
        processCancel(orderId);
        processLimit(std::make_shared<Order>(orderId, side, newPrice, newQuantity,
                                             nowNanos(), tif, expiry));
    }
}

int OrderBook::expireOrders(long currentTimeNanos) {
    std::vector<long> toExpire;
    for (const auto& [id, order] : orderMap_) {
        if (order->timeInForce() == TimeInForce::GTD &&
            order->expiryTimestamp() <= currentTimeNanos) {
            toExpire.push_back(id);
        }
    }
    for (long id : toExpire) processCancel(id);
    return static_cast<int>(toExpire.size());
}

long OrderBook::totalAskQuantity() const {
    long total = 0;
    for (const auto& [price, level] : asks_) total += level.totalQuantity();
    return total;
}

long OrderBook::totalBidQuantity() const {
    long total = 0;
    for (const auto& [price, level] : bids_) total += level.totalQuantity();
    return total;
}

long OrderBook::totalAskQuantityUpTo(long price) const {
    long total = 0;
    for (const auto& [levelPrice, level] : asks_) {  // ascending
        if (levelPrice > price) break;
        total += level.totalQuantity();
    }
    return total;
}

long OrderBook::totalBidQuantityDownTo(long price) const {
    long total = 0;
    for (const auto& [levelPrice, level] : bids_) {  // descending
        if (levelPrice < price) break;
        total += level.totalQuantity();
    }
    return total;
}

TopOfBook OrderBook::getTopOfBook() const {
    long bidPrice = std::numeric_limits<long>::min(), bidQty = 0;
    long askPrice = std::numeric_limits<long>::max(), askQty = 0;

    if (!bids_.empty()) {
        const auto& e = *bids_.begin();
        bidPrice = e.first;
        bidQty = e.second.totalQuantity();
    }
    if (!asks_.empty()) {
        const auto& e = *asks_.begin();
        askPrice = e.first;
        askQty = e.second.totalQuantity();
    }
    return TopOfBook{bidPrice, askPrice, bidQty, askQty};
}

std::vector<BookDepthEntry> OrderBook::getBidDepth() const {
    std::vector<BookDepthEntry> result;
    result.reserve(bids_.size());
    for (const auto& [price, level] : bids_) {
        result.push_back(BookDepthEntry{price, level.totalQuantity(), level.size()});
    }
    return result;
}

std::vector<BookDepthEntry> OrderBook::getAskDepth() const {
    std::vector<BookDepthEntry> result;
    result.reserve(asks_.size());
    for (const auto& [price, level] : asks_) {
        result.push_back(BookDepthEntry{price, level.totalQuantity(), level.size()});
    }
    return result;
}

}
